server/datastore/mysql/scripts.go

```diff
@@ -569,3 +580,72 @@ func (ds *Datastore) GetHostLockWipeStatus(ctx context.Context, hostID uint, fle
 	}
 	return status, nil
 }
+
+// LockHost will create the script execution request and update host_mdm_actions in a single transaction.
+func (ds *Datastore) LockHostViaScript(ctx context.Context, request *fleet.HostScriptRequestPayload) error {
+	var res *fleet.HostScriptResult
+	return ds.withRetryTxx(ctx, func(tx sqlx.ExtContext) error {
+		var err error
+		res, err = newHostScriptExecutionRequest(ctx, request, tx)
+		if err != nil {
+			return ctxerr.Wrap(ctx, err, "lock host via script create execution")
+		}
+
+		const stmt = `
+	INSERT INTO host_mdm_actions
+	(
+		host_id,
+		lock_ref,
+		unlock_ref
+	)
+	VALUES (?,?,NULL)
+	ON DUPLICATE KEY UPDATE
+		lock_ref = VALUES(lock_ref),
+		unlock_ref = NULL
+	`
+
+		_, err = tx.ExecContext(ctx, stmt,
+			request.HostID,
+			res.ExecutionID,
+		)
+		if err != nil {
+			return ctxerr.Wrap(ctx, err, "lock host via script update mdm actions")
+		}
+
+		return nil
+	})
+}
+
+func (ds *Datastore) UnlockHostViaScript(ctx context.Context, request *fleet.HostScriptRequestPayload) error {
+	var res *fleet.HostScriptResult
+	return ds.withRetryTxx(ctx, func(tx sqlx.ExtContext) error {
+		var err error
+		res, err = newHostScriptExecutionRequest(ctx, request, tx)
+		if err != nil {
+			return ctxerr.Wrap(ctx, err, "unlock host via script create execution")
+		}
+
+		const stmt = `
+	INSERT INTO host_mdm_actions
+	(
+		host_id,
+		unlock_ref,
+		lock_ref
+	)
+	VALUES (?,?,NULL)
+	ON DUPLICATE KEY UPDATE
+		unlock_ref = VALUES(unlock_ref),
+		lock_ref = NULL
```

Comment:
Same approach here, I think it'd be best to nullify the other fields (wipe_ref, unlock_pin) on duplicate.

Reply:
@roperzh That's probably the focus of my next PR (I'm working to get the big one y'all have been working out of to be ready to merge in the feature branch, and then move to that). I'll do the script-focused part (so, Windows/Linux), but will try to have a Datastore thingy that can apply that logic for the MDM command too, if that makes sense.